### PR TITLE
Add Stackdriver Trace exporter for trace.

### PR DESCRIPTION
**Pull request by ymotongpoo**
This change adds the exporter to Stackdriver Trace tentatively to the repo.

TODOs for future work is:

* to replace bundler.Bundler
* to add proper tests for the exporter
* to move the exporter to proper repository once it will be created.

### CODE REVIEW

File: exporter/trace/stackdriver/stackdriver.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+	traceapi "cloud.google.com/go/trace/apiv2"
+	"go.opentelemetry.io/api/key"
+	"go.opentelemetry.io/sdk/trace"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+)
+
+// Options contains options for configuring the exporter.
+type Options struct {
```
Comment (dmathieu) [inline exporter/trace/stackdriver/stackdriver.go]: With the Jaeger exporter moving to functional options (#161), how about doing the same here?

File: exporter/trace/stackdriver/stackdriver.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	// to Stackdriver Monitoring. This is only used for Proto metrics export
+	// for now. The minimum number of workers is 1.
+	NumberOfWorkers int
+}
+
+const defaultTimeout = 5 * time.Second
+
+// Exporter is a trace exporter that uploads data to Stackdriver.
+//
+// TODO(yoshifumi): add a metrics exoprter once the spec definition
```
Comment (dmathieu) [inline exporter/trace/stackdriver/stackdriver.go]: Typo: exoprter

File: exporter/trace/stackdriver/stackdriver.go (hunk 1 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Thank you for the suggestion. I changed the code to use functions instead of passing struct directly to the NewExporter function.

File: exporter/trace/stackdriver/stackdriver.go (hunk 2 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Thank you for catching. Fixed!

File: exporter/trace/stackdriver/trace.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		default:
+			log.Printf("OpenTelemetry Stackdriver exporter: failed to upload %d spans: buffer full", o.accum)
+			o.accum = 0
+			o.delay()
+		}
+	})
+}
+
+func (o *overflowLogger) log() {
+	o.mu.Lock()
```
Comment (jmacd) [inline exporter/trace/stackdriver/trace.go]: It looks dangerous to `log.Println` while holding a lock, I'd say.

File: exporter/trace/stackdriver/trace.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	if err != nil {
+		// TODO(ymotongpoo): handle detailed error categories
+		// span.SetStatus(codes.Unknown)
+		e.o.handleError(err)
+	}
+}
+
+// overflowLogger ensures that at most one overflow error log message is
+// written every 5 seconds.
+type overflowLogger struct {
```
Comment (jmacd) [inline exporter/trace/stackdriver/trace.go]: I believe we've discussed handling errors of this nature in a more holistic manner. Every bit of code needs to pass errors back to the user, and not all users will want these on the console. We have discussed registering a callback with the SDK as a means of conveying errors.

File: example/http-stackdriver/server/modd.conf [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# A basic modd.conf file for Go development.
+
+# Run go test on ALL modules on startup, and subsequently only on modules
+# containing changes.
+server.go {
+    daemon +sigterm: go run server.go
+}
```
Comment (rghetia) [inline example/http-stackdriver/server/modd.conf]: add newline.

File: exporter/trace/stackdriver/stackdriver.go [hunk 6]
```diff
@@ -0,0 +1,258 @@
+// Copyright 2019, OpenTelemetry Authors
+// Copyright 2017, OpenCensus Authors
```
Comment (rghetia) [inline exporter/trace/stackdriver/stackdriver.go]: multiple copyright statements. OR is it deliberately added to acknowledge that it is based from opencensus stackdriver exporter?

File: exporter/trace/stackdriver/stackdriver.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+		timeout = defaultTimeout
+	}
+	return context.WithTimeout(ctx, timeout)
+}
+
+// Register registers e to the tracer.
+func (e *Exporter) Register() {
+	trace.Register()
+	trace.RegisterExporter(e)
+}
```
Comment (rghetia) [inline exporter/trace/stackdriver/stackdriver.go]: trace.Register() should be called by application.
RegisterExporter will no longer be there after #162 is merged.
Exporter should be used with either SimpleSpanProcessor or BatchSpanProcessor. BatchSpanProcessor would be more appropriate.

File: exporter/trace/stackdriver/stackdriver.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+	return &newSD
+}
+
+// Flush waits for exported data to be uploaded.
+//
+// This is useful if your program is ending and you do not
+// want to lose recent stats or spans.
+func (e *Exporter) Flush() {
+	e.traceExporter.Flush()
+}
```
Comment (rghetia) [inline exporter/trace/stackdriver/stackdriver.go]: If BatchSpanProcessor (bsp) is used then Flush() is not required. Shutdown() call in bsp is equivalent.

File: exporter/trace/stackdriver/trace.go [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+func newTraceExporterWithClient(o *options, c *traceclient.Client) *traceExporter {
+	e := &traceExporter{
+		projectID: o.ProjectID,
+		client:    c,
+		o:         o,
+	}
+	b := bundler.NewBundler((*tracepb.Span)(nil), func(bundle interface{}) {
+		e.uploadFn(bundle.([]*tracepb.Span))
+	})
```
Comment (rghetia) [inline exporter/trace/stackdriver/trace.go]: bundler can be removed with BatchSpanProcessor.

File: example/http-stackdriver/server/modd.conf (hunk 5 above)
Comment (ymotongpoo) [reply example/http-stackdriver/server/modd.conf]: Done

File: exporter/trace/stackdriver/stackdriver.go (hunk 6 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: I meant to put multiple copyright owners because most of codes are from OpenCensus Stackdriver Exporter.
https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/blob/master/stackdriver.go

File: exporter/trace/stackdriver/stackdriver.go (hunk 7 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Done

File: exporter/trace/stackdriver/stackdriver.go (hunk 8 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: ACK. Changed the method name from Flush to Shutdown.
https://godoc.org/go.opentelemetry.io/sdk/trace#BatchSpanProcessor.Shutdown
https://godoc.org/go.opentelemetry.io/sdk/trace#SimpleSpanProcessor.Shutdown

File: exporter/trace/stackdriver/trace.go (hunk 9 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/trace.go]: In that case, my understanding is that ExportSpans need to implement all implementations to upload SpanData in batch safely. Is that correct? It seems Jaeger exporter is still using bundler for SimpleSpanProcessor.
https://github.com/open-telemetry/opentelemetry-go/blob/master/exporter/trace/jaeger/jaeger.go#L104-L108

Comment (rghetia) [reply exporter/trace/stackdriver/trace.go]: Jaeger should be updated to use BatchSpanProcessor as well.

File: exporter/trace/stackdriver/trace.go (hunk 3 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/trace.go]: Eventually, overflowLogger was totally removed as a result of elimination of bundler.

File: exporter/trace/stackdriver/trace.go (hunk 4 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/trace.go]: Eventually, overflowLogger was totally removed as a result of elimination of bundler.

File: exporter/trace/stackdriver/trace.go (hunk 9 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/trace.go]: google.golang.org/api/support/bundler was totally removed.

File: exporter/trace/stackdriver/stackdriver.go [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+	if err != nil {
+		return err
+	}
+	trace.RegisterSpanProcessor(spanProcessor)
+	return nil
+}
+
+// RegisterSimpleSpanProcessor registers e as SimpleSpanProcessor.
+func (e *Exporter) RegisterSimpleSpanProcessor() {
+	spanProcessor := trace.NewSimpleSpanProcessor(e)
```
Comment (rghetia) [inline exporter/trace/stackdriver/stackdriver.go]: this creates a new local spanProcessor object and does not update the global one. Anyway, once you move spanProcess inside Exporter you don'e need the global one.

File: exporter/trace/stackdriver/stackdriver.go [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+func (e *Exporter) ExportSpan(sd *trace.SpanData) {
+	if len(e.traceExporter.o.DefaultTraceAttributes) > 0 {
+		sd = e.sdWithDefaultTraceAttributes(sd)
+	}
+	e.traceExporter.ExportSpan(sd)
+}
+
+// ExportSpans exports a slice of SpanData to Stackdriver Trace in batch
+func (e *Exporter) ExportSpans(sds []*trace.SpanData) {
+	// TODO(ymotongpoo): implement here
```
Comment (rghetia) [inline exporter/trace/stackdriver/stackdriver.go]: Please remove the comment.

File: exporter/trace/stackdriver/trace.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+	}
+	e.uploadFn(pbSpans)
+}
+
+// Shutdown waits for exported trace spans to be uploaded.
+//
+// This is useful if your program is ending and you do not want to lose recent
+// spans.
+func (e *traceExporter) Shutdown() {
+	spanProcessor.Shutdown()
```
Comment (rghetia) [inline exporter/trace/stackdriver/trace.go]: spanProcessor should be unregistered instead of calling Shutdown directly. When SDK implements Shutdown this will not be necessary.
Not doing Unregister would still allow SDK to call OnStart and OnEnd on spanProcessor which in turn will call ExportSpan or ExportSpans.

File: exporter/trace/stackdriver/trace.go [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+func (e *traceExporter) OnStart(sd *trace.SpanData) {
+	spanProcessor.OnStart(sd)
+}
+
+func (e *traceExporter) OnEnd(sd *trace.SpanData) {
+	spanProcessor.OnEnd(sd)
+}
```
Comment (rghetia) [inline exporter/trace/stackdriver/trace.go]: What is the use of OnStart and OnEnd here?

File: exporter/trace/stackdriver/trace.go [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+// uploadSpans uploads a set of spans to Stackdriver.
```
Comment (rghetia) [inline exporter/trace/stackdriver/trace.go]: It stutters, maybe rephrase to "uploadSpans sends a set of spans to Stackdriver".

File: exporter/trace/stackdriver/stackdriver.go (hunk 6 above)
Comment (rghetia) [reply exporter/trace/stackdriver/stackdriver.go]: I think putting just one copyright with OpenTelemetry Authors is fine. We will have a reference to OpenCensus in main License file.

File: exporter/trace/stackdriver/stackdriver.go [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	if timeout <= 0 {
+		timeout = defaultTimeout
+	}
+	return context.WithTimeout(ctx, timeout)
+}
+
+var spanProcessor trace.SpanProcessor
```
[the remark this answers is not in the document]
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Done

File: exporter/trace/stackdriver/stackdriver.go [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+	queueSizeThreshold := 50
+	if e.traceExporter.o.QueueSizeThreshold > 0 {
+		queueSizeThreshold = e.traceExporter.o.QueueSizeThreshold
+	}
+
+	spanProcessor, err := trace.NewBatchSpanProcessor(e,
+		trace.WithMaxExportBatchSize(queueSizeThreshold),
+		trace.WithMaxQueueSize(queueSizeThreshold),
+		trace.WithScheduleDelayMillis(delayThreshold),
+	)
```
[the remark this answers is not in the document]
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Totally makes sense. Fixed as suggested.

File: exporter/trace/stackdriver/stackdriver.go (hunk 10 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Done with the fix that move spanProcessor to Exporter field.

File: exporter/trace/stackdriver/stackdriver.go (hunk 11 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Done.

File: exporter/trace/stackdriver/trace.go (hunk 12 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/trace.go]: Because I moved spanProcessor to Exporter, I removed Shutdown from traceExporter and implemented Shutdown as Exporter's method as you suggested.

File: exporter/trace/stackdriver/trace.go (hunk 13 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/trace.go]: As there's Shutdown, I thought it should have OnStart and OnEnd to follow, but I found these were not called from Exporter. Anyway, I removed these 2 methods from traceExporter because Exporter's Shutdown is not calling SpanProcessor's Shutdown.

File: exporter/trace/stackdriver/trace.go (hunk 14 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/trace.go]: Done.

File: exporter/trace/stackdriver/stackdriver.go (hunk 6 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Removed copyright statements of OpenCensus from license headers.

File: exporter/trace/stackdriver/stackdriver.go [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+	if err != nil {
+		return err
+	}
+	trace.RegisterSpanProcessor(e.spanProcessor)
+	return nil
+}
+
+// RegisterSimpleSpanProcessor registers e as SimpleSpanProcessor.
+func (e *Exporter) RegisterSimpleSpanProcessor() {
+	spanProcessor := trace.NewSimpleSpanProcessor(e)
```
Comment (rghetia) [inline exporter/trace/stackdriver/stackdriver.go]: this is assigned to local spanProcessor and not e.spanProcessor.

I think RegisterSimpleSpanProcessor is not necessary.

File: exporter/trace/stackdriver/stackdriver.go [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+	}
+	if e.traceExporter.o.MaxQueueSize > 0 {
+		opts = append(opts, trace.WithMaxQueueSize(e.traceExporter.o.MaxQueueSize))
+	}
+	if e.traceExporter.o.MaxExportBatchSize > 0 {
+		opts = append(opts, trace.WithMaxExportBatchSize(e.traceExporter.o.MaxExportBatchSize))
+	}
+
+	var err error
+	e.spanProcessor, err = trace.NewBatchSpanProcessor(e, opts...)
```
Comment (rghetia) [inline exporter/trace/stackdriver/stackdriver.go]: This should be protected using a lock and should be allowed only once.

File: exporter/trace/stackdriver/stackdriver.go [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+		}
+	}
+	newSD.Attributes = append(newSD.Attributes, sd.Attributes...)
+	return &newSD
+}
+
+// Shutdown unregisters spanProcessor.
+func (e *Exporter) Shutdown() {
+	e.spanProcessor = nil
+}
```
Comment (rghetia) [inline exporter/trace/stackdriver/stackdriver.go]: Did you forget to change this to unregister?

File: exporter/trace/stackdriver/stackdriver.go (hunk 17 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Oops. I missed the update here. Fixed.

File: exporter/trace/stackdriver/stackdriver.go (hunk 19 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: I just realized the function. Added the process before the line.

File: exporter/trace/stackdriver/stackdriver.go (hunk 18 above)
Comment (ymotongpoo) [reply exporter/trace/stackdriver/stackdriver.go]: Embedded sync.Once in Exporter to lock and do the initialization only once.
